### DISCUSSION TITLE
SyntaxError: Unexpected token V in JSON at position 113

### DISCUSSION THREAD

**Discussion by racerx9146** [tagyoureit/nodejs-poolController#463] [Q&A] [answered]
Hello All

1st time setup. I was thinking about purchasing a connected pool controller board but then found this software. I have some experience with running Homebridge under Windows. I wanted to start over with a Raspberry Pi but can't find one for sale right now. So i have attempted to add the poolController to the same Windows environment.

I got my protocol adapter working. I have tested under Windows with RealTerm and i can see the correct hex.

I have done the install but after I start I get this error

> nodejs-poolcontroller@7.7.0 build
> tsc

Error reading configuration information.  Aborting startup: SyntaxError: Unexpected token V in JSON at position 113

C:\Users\Daron\nodejs-poolController\config\Config.ts:38
            const def = JSON.parse(fs.readFileSync(path.join(process.cwd(), "/defaultConfig.json"), "utf8").trim());
                             ^
SyntaxError: Unexpected token V in JSON at position 113
    at JSON.parse (<anonymous>)
    at new Config (C:\Users\Daron\nodejs-poolController\config\Config.ts:38:30)
    at Object.<anonymous> (C:\Users\Daron\nodejs-poolController\config\Config.ts:191:31)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (C:\Users\Daron\nodejs-poolController\logger\Logger.ts:23:1)

Not quite sure what this means. My versions
Node.js  v16.13.0
NPM v8.8.0

regards Daron

**Comment by rstrouse** [comment]
The defaultConfig.json file is corrupt.  This happens if you manually edit this file and missed a comma, quote, or bracket.  It looks like you may have added some text to it that has a V in it.  However, you should never edit this file.  The defaultConfig.json file is used to provide the initial settings for the software.   All persistent changes for your installation will end up in a config.json file that is generated on first run of the application.  Even then you shouldn't need to manually edit these json files.  Almost all configuration for njsPC can be done from dashPanel.

**Reply by racerx9146** [reply]
appreciate your reply very much. I thought you may be onto something so I deleted my entire directory and reinstalled from scratch per the instructions. I can then start the node but it errors trying to open the com port. Then if I edit the config.json for my actual com port and the rs485Port i get the error I initially started the post with. I am using the Elfin EW11 protocol adapter mentioned in another thread.

Below is the except from my config.json. When I add the value FABULATECH\VSPC\DEVICE4294967294 it when it all goes wonky :)

{
  "controller": {
    "comms": {
      "portId": 9,
      "enabled": true,
      "rs485Port": "FABULATECH\VSPC\DEVICE4294967294",
      "mockPort": false,
      "netConnect": false,
      "netHost": "raspberrypi",
      "netPort": 9801,
      "inactivityRetry": 10,
      "portSettings": {
        "baudRate": 9600,
        "dataBits": 8,
        "parity": "none",
        "stopBits": 1,
        "flowControl": false,
        "autoOpen": false,
        "lock": false
      }

**Reply by racerx9146** [reply]
update... i changed my setting for rs485port to the "virtual path" the Elfin/InternetofThings software creates. Now when I start up it tries but I get this

5/27/2022, 10:36:06 AM] warn: Inactivity timeout for 9 serial port PENTAIR/Connect after 10 seconds
[5/27/2022, 10:36:06 AM] error: Error opening port 9: Opening PENTAIR/Connect: Unknown error code 3. Retry in 10 seconds
[5/27/2022, 10:36:16 AM] warn: Inactivity timeout for 9 serial port PENTAIR/Connect after 10 seconds
[5/27/2022, 10:36:16 AM] error: Error opening port 9: Opening PENTAIR/Connect: Unknown error code 3. Retry in 10 seconds
        [5/27/2022, 10:36:26 AM] warn: Inactivity timeout for 9 serial port PENTAIR/Connect after 10 seconds
[5/27/2022, 10:36:26 AM] error: Error opening port 9: Opening PENTAIR/Connect: Unknown error code 3. Retry in 10 seconds
[5/27/2022, 10:36:36 AM] warn: Inactivity timeout for 9 serial port PENTAIR/Connect after 10 seconds
[5/27/2022, 10:36:36 AM] error: Error opening port 9: Opening PENTAIR/Connect: Unknown error code 3. Retry in 10 seconds
[5/27/2022, 10:36:46 AM] warn: Inactivity timeout for 9 serial port PENTAIR/Connect after 10 seconds
[5/27/2022, 10:36:46 AM] error: Error opening port 9: Opening PENTAIR/Connect: Unknown error code 3. Retry in 10 seconds

**Comment by tagyoureit** [comment, marked as answer]
First off, we don't have great integration with HomeBridge.  Someone would need to step up and build that integration.  Just want to throw it out there.  An integration could be native, MQTT, etc.  You could also bridge it through Hubitat (eg [bridge](https://github.com/tonesto7/homebridge-hubitat-tonesto7)) if you have one.

On the Elfin, it can be configured for direct access or as a NetConnect server.  How do you have it setup?  If it's local, you should be able to select the port (screenshot from a Pi, but idea is the same) from the dropdown.
<img width="631" alt="2022-05-27_12-16-13" src="https://user-images.githubusercontent.com/1646840/170776688-ddd2cae3-a12d-48ba-9b69-a36bf9b02a4f.png">

If you are using the Elfin server, you need to select port type = Network and put in the IP.  Again, this can all be done from the UI.

**Reply by rstrouse** [reply]
You were adding backslashes into a json file.  Please use the UI and you will see the difference.  The \V is the reason why you were blowing it up it sees V as being invalid since it expected an escape character right after it.   This in fact should probably be a forward slash / if it is meant to be a URI.  So like @tagyoureit said get njsPC and dashPanel running and use the UI to make your changes.  It will properly escape things like slashes and keep the config.json file from being corrupted.

If you are seeing the log disconnecting and reconnecting this is because there is either no RS485 equipment on the bus or you have not set any equipment up to communicate with.   If there is silence njsPC will restart the connection to ensure there are no hangs and this is normal.  I am guessing that IoT address is an MQTT protocol address not a TCP address.  Simply put the ip address and port of the adapter in after you select `Network` from the dropdown.  The port type is not local and should be a network port.

**Comment by racerx9146** [comment]
Many thanks for all the help and being patient with me. I have some progress. 

I was working on the assumption I should get the pool controller working perfect before firing up the dash panel. I got the dash panel loaded and was able to delete my attempt at manually adding the virtual com and let the dash panel do it. I seem to now have the controller running clean but its not "discovering" my pentair junk :) Even after a restart of everything. 

I was able to see the correct string with terminal emulation so I think my rs485 can at least listen. I know the Elfin can do everything including make you a sandwich but I am just using the software that comes with it to create a virtual serial port locally much the usb-serial port drivers do.
So my connection is Pentair Com Port -> Elfin -> Wifi -> local pc running Elfin software as virtual com 9 for me on the same pc running the poolcontroller node

![Screen Shot 2022-05-27 at 5 06 55 PM](https://user-images.githubusercontent.com/78169155/170802291-dea4faac-2fcb-41fe-bc76-64f98d8f1f4f.png)

![Screen Shot 2022-05-27 at 5 09 00 PM](https://user-images.githubusercontent.com/78169155/170802297-8c22de24-50e0-4b17-9782-b38040968c64.png)

**Reply by rstrouse** [reply]
Are you running a Pentair controller such as EasyTouch, IntelliTouch, or IntelliCenter?  If you are not then you need to go to the equipment and add all the items that you have on the bus.  njsPC does not indiscriminately call out to equipment.   However, if you have one of the Pentair controllers this will notify njsPC of its existence and njsPC will become its slave.  

If you have one of these then you are still not communicating properly on the RS485 bus.  If a Pentair outdoor control panel (OCP) is present then all the installations I have seen with an Elfin adapter simply use a network connection that is tied to the URL of the adapter.

**Reply by racerx9146** [reply]
Success!!!! many thanks to all. Yes I am running a Pentair Easytouch. When i checked the RS485 bus via terminal emulator last night I saw the correct string. After my progress today I checked again and could NOT see the correct string. On the Elfin software I locked down the virtual com port to 9600 baud and all of a sudden I saw the correct string. So yes i suspect there is a easier way to use the Elfin as you suggest above with direct connect to the adapter URL. That way I dont have to run the virtual com port software. But I think its time for a drink! Again thanks to all who help me plod along. I am sure many more questions to come but I will give you a break for a while :)

![Screen Shot 2022-05-27 at 6 03 56 PM](https://user-images.githubusercontent.com/78169155/170803869-f6026c2f-1b00-4bc8-a2f1-bc9305371081.png)

**Reply by rstrouse** [reply]
Its a little bit concerning that EasyTouch thinks that both the pool and the spa are running at the same time for a shared body system and the solar temp sensor is either not installed (and you have the solar is present checked) or it is shorted out.

**Reply by tagyoureit** [reply]
There are a few odd things in your picture you should keep an eye on...
1. Your solar temp should not be 252 degrees.
2. Your pool and spa should never both be on at the same time
3. Your heat mode is off for both the pool and spa, but it still thinks the heater is on.  Again, the pool/spa heater should never be on at the same time.
4. Are the schedules right?  Do you only have a single pool schedule from 1-4:30am?
5. You are missing any pumps that EasyTouch knows about.  You should have a primary pump listed that ET can control.

Weird, wild stuff...

**Comment by racerx9146** [comment]
Yeah Its getting a bit smoother

1) I dont have solar
2) This has resolved
3) resolved
4) schedule is right but I will be fine tuning more
5) I was able to add chlorinator but my pump is undefined. I have a Intelliflow 4-160, when I try to configure it it wont stay. not sure what choice the 4-160 maps to in the drop down

**Reply by rstrouse** [reply]
4-160 is an IntelliFlo VS pump (the original).  I too have two of these.

**Reply by racerx9146** [reply]
Many thanks for all the help. I have my setup working quite well. Took a bit of messing around with the settings to get the pool pump defined but its fun. Very responsive interface. The Homebridge plug-in does not seem to work for me but you did warn me. That's a nice to have versus a must have anyway. 

I do have another question, I have 2 Jandy 2 position valves just for switching the flow from the pool body or water to the spa which both run off the intelliflow pump. My instance came with 4 defined. Is right and is there any purpose in defining them further? Does the Easy touch "tell" the poolController of valve commands?

regards Daron
![Screen Shot 2022-06-01 at 7 33 46 AM](https://user-images.githubusercontent.com/78169155/171430100-e89f7eed-13a6-4352-8125-a138774c0f4c.png)

**Reply by tagyoureit** [reply]
You should only have the intake and return.  Valve A is typically for a solar heater.  But if you only have a pool/spa than Valve A should not be defined as a Spa valve.  The intake/return take care of pool, spa, and spillway functions.  Did you set the Valve A as Spa or did it come that way?  I'd remove the spa function from that valve as it isn't doing anything.

Easytouch doesn't communicate the valve status.  It's almost impossible to know because each valve has 3 settings and the valves are either directed to spin one way or spin the other way depending on which wires are hot.

**Reply by rstrouse** [reply]
The 4 valves that you see in this list are hardwired on the EasyTouch2 8.  On the personality board there will be an Intake, Return, Valve A, and Valve B connectors that can be used to control 24vac valve actuators.  On a shared body only the Intake and Return valves are required.  The other two are optional.  The way this works is that when the spa body is turned on the intake and return valves rotate to their diverted position and when it is turned off they rotate back to the default position.  In the case of ET28 these two valve assignments are hardcoded to an internal Pool/Spa circuit.  

Valves A&B are user definable and can be assigned to any circuit with couple of exceptions.  If you have solar installed Valve A must be used for the solar valve.   If you don't it will not work properly with the delays.   Pentair has something hardcoded in their firmware to only work with valve A wrt solar delay. 

Since you have a shared body system you have an intake and return plus a Valve A&B.  Since I see that valve A has been redefined as Spa, I can only assume that is being used as a heater bypass.  I can't think of any other purpose to assign that valve to the Spa circuit.  If you do not have 3 valve actuators on the pad then this is simply doing nothing and does not matter what it is assigned to.  The gist of this is that when the circuit that is assigned to the valve is turned on, the valve will rotate to the diverted position.

@tagyoureit filled you in on state information while I typed out this dissertation.
